lib/recursive-methods.rb

```diff
@@ -1,49 +1,117 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n == 0 || n == 1
+    return 1
+  elsif n < 0
+    raise ArgumentError
+  end
+  
+  until n == 1 do
+    return n * factorial(n-1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length <= 1
+    return s
+  end
+  
+  reversed_str = reverse(s[1..-1])
+  reversed_str += s[0]
+  reversed_str
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length <= 1
+    return s
+  end
+  
+  return s[s.length-1] + reverse_inplace(s[0, s.length-1])
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  if n == 0
+    return 0
+  elsif n == 1
+    return 2
+  else
+    return 2 + bunny(n - 1)
+  end
+  
+  return
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length.odd?
+    return false
+  elsif
+    s.length == 0
+    return true
+  end
+  if s[0] != s[-1]
+    return nested(s[1..-2])
+  else
+    return false
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(1)
+
+def search(array,value)
```

Comment:
👍 This works, but you have similar time/space issues with the above methods due to deleting elements.